### DISCUSSION TITLE
Is DisplayInformation no longer supported?

### DISCUSSION THREAD

**Discussion by ThomasArdal** [unoplatform/uno#6012] [Q&A] [answered]
After having upgraded my project to `Uno.UI` version `3.7.3` the following code causes my application to crash:

```csharp
var displayInfo = DisplayInformation.GetForCurrentView();
if (displayInfo != null)
{
    var orientation = displayInfo.CurrentOrientation; // <-- the code doesn't fail if I comment out this line
}
```

This worked fine previously but now crashes as soon as I use any of the properties on `displayInfo`.

Does anyone know what the problem could be there?

**Comment by pkar70** [comment]
For what platform? iOS/macOS/Android/WASM/... ?

**Reply by ThomasArdal** [reply]
Not sure if all platforms crash. But currently, I'm testing on Android 10.0 (Q) - the emulator.

**Reply by ThomasArdal** [reply]
It could be this error:

```
JNI DETECTED ERROR IN APPLICATION: JNI GetStringUTFChars called with pending exception java.lang.StackOverflowError: stack size 8MB
  at java.lang.String java.lang.Class.getName() (Class.java:742)
  at void crc64122dcf5ad656876d.NativeApplication.n_onCreate() (NativeApplication.java:-2)
  at void crc64122dcf5ad656876d.NativeApplication.onCreate() (NativeApplication.java:26)
  at void android.app.Instrumentation.callApplicationOnCreate(android.app.Application) (Instrumentation.java:1154)
  at void android.app.ActivityThread.handleBindApplication(android.app.ActivityThread$AppBindData) (ActivityThread.java:5871)
  at void android.app.ActivityThread.access$1100(android.app.ActivityThread, android.app.ActivityThread$AppBindData) (ActivityThread.java:199)
 05-15 11:59:05.852 F/ah.Io.UnoSampl( 7536): runtime.cc:566]   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
  at void android.os.Looper.loop() (Looper.java:193)
  at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6669)
  at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
  at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:493)

    in call to GetStringUTFChars
    from void crc64122dcf5ad656876d.NativeApplication.n_onCreate()
```

**Reply by pkar70** [reply]
You upgraded to 3.7.3, but from which version? What is last version your code doesn't crash?
When your code didn't crash, it was also on Android 10 / emulator?
I can correct code of GetCurrentOrientation(), but I don't know if it solves your problem. Current implementation uses deprecated methods, I know how to change it, but I will have some time to do it later, in about 6..8 hours. But maybe problem is somewhere else.

**Reply by ThomasArdal** [reply]
I'm the developer of this sample: https://github.com/elmahio/Elmah.Io.Uno/tree/main/samples/Elmah.Io.UnoSample. The project uses this class library: https://github.com/elmahio/Elmah.Io.Uno/tree/main/src/Elmah.Io.Uno. The old project was running `3.5.1`.

If you are interested in a project reproducing this error, I have pushed it here: https://github.com/elmahio/Elmah.Io.Uno/tree/crash

**Reply by ThomasArdal** [reply]
And BTW, the code that causes the crash is here: https://github.com/elmahio/Elmah.Io.Uno/blob/crash/src/Elmah.Io.Uno/ElmahIoLogger.cs#L80

Looks like this:

```csharp
var width = displayInfo.ScreenWidthInRawPixels;
var height = displayInfo.ScreenHeightInRawPixels;
var orientation = displayInfo.CurrentOrientation;
```

**Reply by pkar70** [reply]
I have to go out, so I have no time just now.
But, if you are developer, you can try yourself this trick (I have no Android 10 device):

Copy relevant code from:
`https://github.com/unoplatform/uno/blob/master/src/Uno.UWP/Graphics/Display/DisplayInformation.Android.cs`
into your app (lines 190..252), and "simulate" GetCurrentOrientation() operation.
If it crash, and if this crash is in line
`windowManager.DefaultDisplay.GetMetrics(displayMetrics);`
then try to change it according to:
`https://stackoverflow.com/questions/63407883/getting-screen-width-on-api-level-30-android-11-getdefaultdisplay-and-getme`

**Reply by ThomasArdal** [reply]
That's where the mystery starts then. I copied the code into the `OnUserInteraction` method on my `MainActivity` class in the Droid project. When I click anywhere on the screen, the method is called and the orientation successfully resolved. So, it seems like the error only happens when I try to access this code from within my class library.

**Reply by pkar70** [reply]
So, it seems like Android Context, required for this code, is inaccessible. Try to check `Android.Content.Context context = Android.App.Application.Context;` in your class library code.

**Reply by ThomasArdal** [reply]
But my library is targeting netstandard why it doesn't have access to that namespace, right? My class library is general and could be included in all of the supported platforms.

**Reply by pkar70** [reply]
it is UI related method, and UI related requires app context. I don't know if Android supports UI methods in something like class library. Check if context is null inside class library.

**Reply by ThomasArdal** [reply]
It would be totally fine with me to not support UI-related methods in class libraries. My point here is that this used to work but didn't work after I upgraded. I can simply remove the failing code from my library and live without the resolution and screen orientation. Just thought you wanted to know that something isn't working as it used to. Could be a decision made at some point, but just hadn't seen it mentioned anywhere.

**Reply by pkar70** [reply]
I will check all updates to DisplayInformation.Android.cs ; all changes between 3.5.1 and 3.7.3.
But crash-related change can be not in Uno but in Xamarin, or some other library used by Uno.

**Reply by pkar70** [reply]
Can you make Issue for this?

**Reply by ThomasArdal** [reply]
Yep, I'll try. Thank you for your help so far.

**Reply by jeromelaban** [reply]
@ThomasArdal Thanks for the repro sample! I'll take a look; having a stackoverflow at that location seems pretty odd.

**Reply by jeromelaban** [reply]
Ok, so the reason this happens is that using `ContextHelper.Current` (directly or indirectly as `DisplayInformation` uses  `ContextHelper.Current`) on Android before the App constructor has finished may fail.

In your case, the Elmah logger is processing a Warning that is generated by a resource being initialized, causing an logging loop (the failure to log causes another log, which fails recursively).

There are multiple ways around this:
- Temporarily use `ContextHelper.Current` to skip uing `DisplayInformation` if null:
```
#if __ANDROID__
if(Uno.UI.ContextHelpher != null)
{
   ...
}
#endif
```
- Handle exceptions for those methods, skip logging it and return default values.
- Wait for https://github.com/unoplatform/uno/issues/6019 to fix this; we'll track this to get the incorrect behavior.

**Reply by jeromelaban** [reply]
One additional detail: The `DisplayInformation` class is only available from the UI thread, and fails with an exception when used on other threads on Windows. The change in https://github.com/unoplatform/uno/pull/6021 will use a similar pattern, but with an exception being raised when used too early in the app.

**Reply by ThomasArdal** [reply]
Thank you for creating the issue, @jeromelaban.

Just a few comments. The `__ANDROID__` constant isn't available from class libraries as I understand it. Or am I wrong there?

Wrapping the code in try/catch doesn't help since it's a stack overflow exception that causes the entire thing to crash.

Thank you so much for looking into this.

**Reply by ThomasArdal** [reply]
And in addition to your sample. Should look like this, right?

```csharp
if (Uno.UI.ContextHelper.Current != null)
{
    ...
}
```

**Reply by jeromelaban** [reply]
The library you've built is a cross-targeted library, which builds for `monoandroid10.0`, making `__ANDROID__` available as well.

That said, it makes sense that the stack overflow would still happen because accessing the `ContextHelper` will log unconditionally if the context is null. This makes me notice that the fix I made for this will not fix your issue, as the log will still happen. I'll adjust that.

**Reply by ThomasArdal** [reply]
I'm not sure I understand the workaround specified in a previous comment:

```csharp
#if __ANDROID__
if(Uno.UI.ContextHelper != null)
{
   ...
}
#endif
```

That code doesn't compile since `ContextHelper` is a class and I cannot check a class for `null`. I was suggesting using `ContextHelper.Current` but that doesn't seem to work either.

**Reply by jeromelaban** [reply]
Yes, that's correct, it'd be 

```csharp
#if __ANDROID__
if(Uno.UI.ContextHelper.Current != null)
{
   ...
}
#endif
```

but it'd fail anyways because the getter logs a warning. This will be fixed by the associated PR.

**Comment by ThomasArdal** [comment, marked as answer]
I can verify that it doesn't work. If you're interested I have updated the sample: https://github.com/elmahio/Elmah.Io.Uno/blob/crash/src/Elmah.Io.Uno/ElmahIoLogger.cs#L78

No worries, I'll wait for the bugfix, I believe.

**Reply by jeromelaban** [reply]
we've updated Uno latest dev (in https://github.com/unoplatform/uno/pull/6021) with a more informative error message. It will still throw an exception (like UWP does) when called too early, but this is manageable on your end when invoking `var displayInfo = DisplayInformation.GetForCurrentView();`.

This will be included in soon to be released Uno 3.8.

**Reply by ThomasArdal** [reply]
Is this not in the newest version of `Uno.UI` on NuGet? I upgraded to `3.8.0-dev.302` and include the following in my code:

```csharp
#if __ANDROID__
if (UnoUI.ContextHelper.Current != null)
{
#endif
try
{
    var displayInfo = DisplayInformation.GetForCurrentView();
    // ...
}
catch {}
#if __ANDROID__
}
#endif
```

The app still crashes on startup.

**Reply by jeromelaban** [reply]
You do not need to use the `UnoUI.ContextHelper.Current != null` around.

**Reply by ThomasArdal** [reply]
I see. I'll remove that. But I guess it still crashes?

**Reply by jeromelaban** [reply]
`DisplayInformation.GetForCurrentView();` will always fail when executed too early, which is why you need the try/catch.

The original reason logging failed in a loop is because of `UnoUI.ContextHelper.Current` is logging a failure which causes the loop.

Let us know if this helps.

**Reply by ThomasArdal** [reply]
Sorry, I believe you already explained that to me 🙈 It works perfectly without the `ContextHelper.Current` call. Thank you so much for your help here.

![giphy (2)](https://user-images.githubusercontent.com/563206/120478412-1ee4c600-c3ad-11eb-8d72-d389e9d332dd.gif)
